test: AMM TXE tests

benesjan opened the pull request:
Fixes #10289

Just like my [other recent PR](https://github.com/AztecProtocol/aztec-packages/pull/16451) decided to add TXE tests for AMM.

## On using AI here
The only thing I did manually here is to merge a few test cases because they were too long-running as setting up the initial state is costly. I also told it to not have precise ratio in the second call to `add_liquidity` to test the refund. Other than this AI pretty much managed to nail it. 

### Update on AI
Code quality was not great and had to redo quite a lot of stuff. That could be improved though if we had a proper reference (plenty of the style issues were present in the rest of our codebase).

benesjan commented:
* **#16458** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16458?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16458?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+// Utility function to setup three token contracts (token0, token1, liquidity_token)
+unconstrained fn setup_tokens(
+    env: &mut TestEnvironment,
+    owner: AztecAddress,
```
minter, not owner. Also no reason for this function to take this param, the account should be created here.

benesjan replied:
Setup functions reworked and fixed the naming 👍

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+        env.create_contract_account()
+    } else {
+        env.create_light_account()
+    };
```
See the other PR for comments on this boolean and why owner needs to authwits.

benesjan replied:
Addressed

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+        env.create_light_account()
+    };
+
+    let (token0_address, token1_address, liquidity_token_address) = setup_tokens(&mut env, owner);
```
I'd consider just inlining this? Or if anything make it so creating a token is a one liner. This a bit of a strange division of scope.

benesjan replied:
Done 👍

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+}
+
+// Utility function to mint tokens to private balance
+unconstrained fn mint_to_private(
+    env: TestEnvironment,
+    token_address: AztecAddress,
+    owner: AztecAddress,
+    to: AztecAddress,
+    amount: u128,
+) {
+    env.call_private(owner, Token::at(token_address).mint_to_private(to, amount));
+}
+
+// Utility function to check private balance
+unconstrained fn check_private_balance(
+    env: TestEnvironment,
+    token_address: AztecAddress,
+    address: AztecAddress,
+    expected_amount: u128,
+) {
+    assert_eq(
+        env.simulate_utility(Token::at(token_address).balance_of_private(address)),
+        expected_amount,
+    );
+}
+
+// Utility function to check public balance
+unconstrained fn check_public_balance(
+    env: TestEnvironment,
+    token_address: AztecAddress,
+    address: AztecAddress,
+    expected_amount: u128,
+) {
+    assert_eq(
+        env.view_public(Token::at(token_address).balance_of_public(address)),
+        expected_amount,
+    );
+}
```
rm

benesjan replied:
Addressed

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+
+// Happy path tests
+// Note: All these tests are quite long-running so I decided to create larger tests (e.g. testing add and remove
+// liquidity happy path just in 1 test case)
```
How long running? Given they run in parallel it's not the end of the world to have long tests, I'm not sure we should take this path so early.

benesjan replied:
It's not too bad but it also makes the tests quite boilerplatey. I would say the result is quite ok but can create a separate test case only for adding liquidity if you would prefer.

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+//    - Transfers liquidity tokens to AMM
+//    - Calls remove_liquidity() to burn tokens and receive underlying assets
+//    - Verifies:
+//      * Owner received correct proportion of tokens back (495 token0, 990 token1)
```
Not a huge fan of this as it can be easy for the description to get out of sync with the test, and then it'll just be confusing. Ideally the test would be easy to follow, and these properties would be evident from reading it. If the test is too long because of excessive boilerplate, then we can use functions to make it smaller.

benesjan replied:
good point. Nuked it and refactored the tests to use helper funcs

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+    env.call_private(
+        owner,
+        AMM::at(amm_address).add_liquidity(initial_amount0, initial_amount1, 0, 0, 1),
+    );
```
The token admin adding liquidity is unnecessarily confusing. It should be a different account.

benesjan replied:
Addressed

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test.nr:
```diff
@@ -0,0 +1,801 @@
+
+    // User2 should have 0 token0 and 200 token1 (refunded excess)
+    check_private_balance(env, token0_address, user2, 0);
+    check_private_balance(env, token1_address, user2, 200);
```
The combination of hardcoded amounts and the numeric variables above will make this annoying to maintain. Ideally we'd either have variables that have been sanely computed, or we'd have all of the values close to one another to be able to verify with relative eas that the thing makes sense.

benesjan replied:
Addressed 👍

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test/test.nr:
```diff
@@ -0,0 +1,594 @@
+global DEFAULT_AMOUNT_0_MIN: u128 = 0;
+global DEFAULT_AMOUNT_1_MIN: u128 = 0;
+
+// Note: Ideally this test would be split into 2 separate test cases - one for order creation and one for order
+// fulfillment, with the second test running on the state from the first test. Since this feature is not currently
+// supported, combining them into a single comprehensive test is the best approach.
```
What do you mean by this? What feature?

benesjan replied:
I was thinking that it would be cool if you could run test case on state of a previous test case. Since here re-running the add liquidity for the test that would test liquidity removal takes what feels like a lot of resources.

Can nuke the comment if you prefer as I don't care.

nventuro commented on noir-projects/noir-contracts/contracts/app/amm_contract/src/test/test.nr:
```diff
@@ -0,0 +1,594 @@
+    assert_eq(initial_liquidity_token_supply, AMM::MINIMUM_LIQUIDITY + AMM::INITIAL_LIQUIDITY);
+
+    // ADDING LIQUIDITY AGAIN
+    let amount0_max = initial_amount0 / 2;
```
```suggestion
    let expected_amount_0_in = initial_amount0 / 2;
```

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/16458/commits/96e84c8df41f893a43c92f5d0dcd39722e37bbf8

nventuro reviewed (approved):
These look much better, thanks. It's still a bit annoying to write these, we'll need to iterate on the style a bit more. Perhaps a struct with scenario config, idk. The fact that we can't return the contract types from functions or put them in structs is very annoying.

benesjan committed:
test: AMM TXE tests

Fixes #10289

Just like my [other recent PR](https://github.com/AztecProtocol/aztec-packages/pull/16451) decided to add TXE tests for AMM.

## On using AI here
The only thing I did manually here is to merge a few test cases because they were too long-running as setting up the initial state is costly. I also told it to not have precise ratio in the second call to `add_liquidity` to test the refund. Other than this AI pretty much managed to nail it.

### Update on AI
Code quality was not great and had to redo quite a lot of stuff. That could be improved though if we had a proper reference (plenty of the style issues were present in the rest of our codebase).
